cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,356 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/case.hpp>
+#include <cudf/utilities/error.hpp>
+#include <strings/char_types/is_flags.h>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace { // anonym.
+  
+  //base class for probing string
+  //manipulation memory load requirements;
+  //
+  struct probe_base
+  {
+    probe_base(column_device_view const d_column,
+               character_flags_table_type const* d_flags,
+               character_cases_table_type const* d_case_table):
+      d_column_(d_column),
+      d_flags_(d_flags),
+      d_case_table_(d_case_table)
+    {  
+    }
+
+    __host__ __device__
+    column_device_view const get_column(void) const
+    {
+      return d_column_;
+    }
+
+    __host__ __device__
+    character_flags_table_type const* get_flags(void) const
+    {
+      return d_flags_;
+    }
+
+    __host__ __device__
+    character_cases_table_type const* get_case_table(void) const
+    {
+      return d_case_table_;
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) const = 0;
+    
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+  };
+
+  
+  //base class for executing string modification:
+  //
+  struct execute_base
+  {
+    execute_base(column_device_view const d_column,
+                 character_flags_table_type const* d_flags,
+                 character_cases_table_type const* d_case_table,
+                 int32_t const* d_offsets,
+                 char* d_chars):
+      d_column_(d_column),
+      d_flags_(d_flags),
+      d_case_table_(d_case_table),
+      d_offsets_(d_offsets),
+      d_chars_(d_chars)
+    {
+    }
+
+    __host__ __device__
+    column_device_view const get_column(void) const
+    {
+      return d_column_;
+    }
+
+    __host__ __device__
+    character_flags_table_type const* get_flags(void) const
+    {
+      return d_flags_;
+    }
+
+    __host__ __device__
+    character_cases_table_type const* get_case_table(void) const
+    {
+      return d_case_table_;
+    }
+
+    __host__ __device__
+    int32_t const* get_offsets(void) const
+    {
+      return d_offsets_;
+    }
+
+    __host__ __device__
+    char* get_chars(void)
+    {
+      return d_chars_;
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) = 0;
```

Comment:
This is never used. Also, it will likely not work here since the virtual-function-table needs to be created in device memory and therefore requires a class of this type be instantiated in a kernel somehow. Because the base class is never referenced in this code, the virtual function never has to be resolved.
I recommend removing this line.

Reply:
It's not _meant_ to be used. It _cannot_ be used, because it's pure virtual. The base class is an abstract class. It's only purpose is to enforce an interface to adhere to. 

The interface specifies how the derived classes are to be constructed and is a common placeholder for the private data members. Instead of duplicating code all over the place, such a base class provides a unique placeholder for the members.

Not providing an interface means having no mechanism to enforce derived classes to adhere to a common interface. They might as well be independent classes. But if they are independent, then the data members must be duplicated all over the place. So to enforce an interface to adhere to you need an abstract class. Which has zero-cost as it _cannot_ be instantiated. Hence the pure virtual function to override. Non-virtual functions _with the same name_ of classes driving from the same base are introducing more problems than they can solve. ("Item 37:  Never redefine an inherited nonvirtual function." Scott Meyers, Effective C++, 2nd ed.; since they're doing different things in the derived classes...they must be redefined; hence they must be virtual!)

Also, there is no runtime cost for this setup, because the virtual table is not used at runtime. And that is because the member function in question is _not_ called via a base pointer or reference. It's called directly via the derived object, so the compiler is optimizing the call.

Abstract classes are the _de facto_ mechanism of enforcing interfaces in C++. As long as no runtime cost is incurred (and, as explained above, it is not) there shouldn't be anything wrong with using C++ abstract classes. After all ...this is a C++ library.

Reply:
I disagree with this. If the programmer does not provide an `operator()` method then this code would not compile without or without this virtual function definition. I feel you are misleading the class user by implying polymorphism which if acted upon would fail at runtime and not compile time.

Reply:
I understand @aschaffer's desire to want to declare the expected interface in the `execute_base` class. I agree that the way it is being used is simply to define an interface and there is no dynamic polymorphism occurring and is therefore safe and correct. 

I am not sure if it is actually _necessary_ to enforce an interface here. If someone were to attempt to use a type that derives from `execute_base` as a function object, then it would _already_ be required to have an `operator()` defined. This in effect already enforces an interface for all derived types.  

I also share @davidwendt's concern by making it pure virtual it could mislead future developers into thinking they could use a `execute_base` as a base class for dynamic polymorphism. 

> Abstract classes are the de facto mechanism of enforcing interfaces in C++. 

Perhaps, but they are not the _only_ way to enforce an interface. You could also use CRTP.

Reply:
The users have no reason to assume polymorphism, when no polymorphic container is involved. But they have the option of doing so, if they want, by involving such a container.

Assume, for a moment, that at some point one would want to apply a `vector` of such functors to a string; or, more realistically, to each row in a `string_view`.

So, one might use (semi-pseudocode-ish):
```
device_vector<probe_base*> functor_v{new probe_capitalize(...), new probe_title(...)}; 
//apply the sequence of functors to string:
for(auto const& fctr: functor_v) {
   fctr(string);
}
```
(assume for a moment that the `operator()` acts directly on `string`'s).

Do you think this code would work _without_ declaring `operator()` _virtual_? (which requires a _pure virtual_ declaration in the base class; which requires an _abstract_ class; and `overridden` versions in derived classes).

The answer is: no, it would not. Reason? Because you have non-virtual classes with same name and signature in derived classes. What you want here is dynamic polymorphism. What you get...is anything but!

**That being said, I will make the change according to your demands. But considered yourselves warned this opens a door for future disasters, in a situation like this. ;-)**

Reply:
Done.